### DISCUSSION TITLE
Limit HTTP Sender Retries

### DISCUSSION THREAD

**Discussion by MPollin0** [nextgenhealthcare/connect#5914] [Q&A]
I have an HTTP sender destination with 3 retires on it. I want to make it so that if a specific error comes as a result of the API call then it will not bother trying the remaining retries.

I have been playing around with response, responseStatus, and responseStatusMessage but nothing I am doing seems to stop all 3 retires from happening. Is this even possible to do?

    var api = response.getMessage()

    if(api.includes('No Data')){
	    responseStatus = SENT   
	    responseStatusMessage = 'SENT'
	    response = SENT
    }

**Comment by rjkroll** [comment]
Are you doing this in your Response transformer or somewhere else?   Setting the responseStatus = SENT should work.   Also, in the past I thought I learned that the response transformer will only run if there is a resposne payload, so for connection problems or other types of server returned errors that contain no payload, nothing will ever get executed in the Response transformer.   You can get around this by setting the Response type to 'Raw' in the Destination/Edit Response settings.

Also, are you sure your IF statement is even going to fire?   You should look into using and parsing the $('responseSatusLine') in the Response Transformer instead of response.GetMessage.    The responseStatusLine will be the "HTTP/1.1 200 OK" ...etc line of the actual resposne message from the transformer.   If you use getMessage, you may not actually be getting anything sometimes.

**Reply by MPollin0** [reply]
Yes, I am doing this in the response transformer. Setting responseStatus = SENT does work in the sense that the response is set to sent, but it is not stopping the HTTP Sender destination from doing the additional retries after, and I have to type set to Raw across the entire channel.


('responseSatusLine') does not entirely help either because the response I am expecting is an error, but an "Okay" error. When someone has "No Data" I would not expect there to be any data there on the additional retries. $('responseSatusLine') will just say ERROR but won't tell me if it is the "No Data" error or an actual other error.

**Reply by pacmano1** [reply]
There is already a send attempts column you can add to the dashboard.

**Comment by pacmano1** [comment]
Well glad you know, maybe you should try using the variable.  

Note: Destination Settings...Queue Messages is set to always.   Response transformer data type is RAW.

```JS
try {
    msg = JSON.parse(msg)                       // we are going to reset msg to be its value in JSON here.
    $c('isJSONresponse', true)
} catch (e) {                               // we have an empty msg or some other non-JSON response 
    $c('isJSONresponse', false)
}

var responseStatusLine = $('responseStatusLine');
var responseCode = parseInt(responseStatusLine.split(' ')[1], 10);

if ( responseCode == 200) {    // we have an entirely successful message
        $c('API_Post_OK',true)
        process_results();   // some function to process the results
} else {                                    // we either have an entirely bad post (endpoint not answering OR we have a JSON error)
    if ($c('isJSONresponse') == true) {             // do not retry, just collect the JSON stuff
        responseStatus = SENT
        // perhaps call another function here
    } else {                                    // Let's retry a few times
        if (connectorMessage.getSendAttempts() <= 3) { // retry a few times
            responseStatus = QUEUED                 
        } else {                                // exceeded the tries
              responseStatus = SENT             // in this we are gonna lie that message sent
              $c('API_Post_OK',false)
              // do something here if needed      
    }
}
```

**Reply by MPollin0** [reply]
Can this work without queueing set to always? I need the destinations in my channel to run one after another

**Reply by pacmano1** [reply]
@MPollin0 , please try it then come back with questions if you run into something you don't expect.
